### ISSUE TITLE
Single expression for combined is and as

### ISSUE THREAD

**Issue by robin-aws** [dafny-lang/dafny#5165]
### Summary

Provide a single expression for casting a value to a type without earlier proving that the value is of that type.

### Background and Motivation

This is a very common pattern in my experience:

```dafny
expect value is SomeTypeWithAVeryLongName;
var casted := value as SomeTypeWithAVeryLongName;
```

### Proposed Feature

Following in the footsteps of constructs like `:- expect/assume`, I propose supporting `value expect as SomeTypeWithAVeryLongTime` and `value assume as SomeTypeWithAVeryLongName`.

**Comment by RustanLeino** [comment]
For the record, here are some similar things that you _can_ write today:

``` dafny
trait Animal { }
class Cat extends Animal { }
class Dog extends Animal { }

method TestForCat(a: Animal) {
  if c: Cat :| c == a {
    print "It's a cat\n";
  } else {
    print "It's something else\n";
  }
}

method SurelyACat(a: Animal)
  requires a is Cat
{
  var c: Cat;
  c :| /*expect*/ c == a;
  print "Yep, was a cat alright\n";
}

method Main() {
  var c := new Cat;
  var d := new Dog;
  TestForCat(c);
  TestForCat(d);
  SurelyACat(c);
}
```

```
% dafny /compile:3 test.dfy
...
Running...

It's a cat
It's something else
Yep, was a cat alright
```

Evidently, we don't support `:| expect`, so I had to write a precondition for `SurelyACat` in order for it to verify.

**Comment by RustanLeino** [comment]
Btw, to more like `:- expect`, I suppose the proposed construct ought to be `as expect`, not `expect as`.

**Comment by robin-aws** [comment]
> Btw, to more like `:- expect`, I suppose the proposed construct ought to be `as expect`, not `expect as`.

Agreed and I started with that, but then felt like `expect as` and `assume as` read a little nicer. :)

**Comment by RustanLeino** [comment]
> > Btw, to more like `:- expect`, I suppose the proposed construct ought to be `as expect`, not `expect as`.
> 
> Agreed and I started with that, but then felt like `expect as` and `assume as` read a little nicer. :)

I agree with that.

**Comment by MikaelMayer** [comment]
I'm really not used to write `expect` statements except for `{:test}` methods, and even less in _expressions_ as you indicate it in the title, so I'm surprised you find this pattern with `expect` common. If you can point me to some examples, I'd be happy to learn more.

I would have assumed that it was common to _test_ rather than _expect_ that a value is of a certain type, because testing can be part of regular expressions.

I did a quick survey on how to achieve something similar:

- You can put the code is/as in a method, but it needs more work as of today as we don't have type bounds. But at least, you could control what is the return value you return
```
method ExpectAs<U, T>(u: U) returns (t: Option<T>) {
  if u is T {
    return Some(u as T);
  } else {
    return None;
  }
}
```

- Use let such that expression, The only failure case we have is `if :| `, and it could be good to think of transforming `if c :| P(c) then X else Y` to `var c :-| P(c);` if P returns a failure-compatible wrapped boolean. 
- Use the `if c: Type :| cond {} else {}` statement or expression
- Use a new pattern matching syntax `match c { case c: Type =>  case _ => ...}`

Is the keyword you mention `x expect as A` meant to mean:
`if x is A then x as A else assume {:axiom} false; match () {}`?
